### ISSUE TITLE
Allow group test functions in Value-Parametrized Tests

### ISSUE THREAD

**Issue by jfinkhaeuser** [google/googletest#2567]
Let's say I have test data like this (looks like in #2566 ? No accident!):

```cpp
struct test_data
{
   foo f;
   bool bar;
} tests[] = {
  { foo(), true },
  { foo(), false },
  // ...
};

class X : public testing::TestWIthParam<test_data>{}

TEST_P(X, first) {}
TEST_P(X, second) {}
```

For the most part, this is exactly what I want. But let's say my `foo` type should behave like a value type, including hashing to sufficiently unique values.

I can always define a standalone test case to do that:

```cpp
TEST(Z, unique_hashes)
{
  std::set<size_t> hashes;
  std::hash<foo> hasher;

  for (size_t i = 0 ; i < sizeof(tests) / sizeof(test_data) ; ++i) {
    hashes.insert(hasher(tests[i].f);
  }

  // The expecation is that the size of the set is the same as the size of the test data
  ASSERT_EQ(sizeof(tests) / sizeof(test_data), hashes.size());
}
```

It works just fine - but the `Z` test is reported as its own unique group when really it belongs to tests on the same test data. I'd like to see something like this:

```cpp
TEST_PG(X, unique_hashes)
{
  std::set<size_t> hashes;
  std::hash<foo> hasher;

  for (auto item : GetValues()) {
    hashes.insert(hasher(item.f);
  }

  // The expecation is that the size of the set is the same as the size of the test data
  ASSERT_EQ(GetValues().size(), hashes.size());
}
```

Which then gets instantiated along with the rest of the suite.

**Comment by derekmauro** [comment]
I'm having trouble understanding what value this provides over just writing a non-parameterized test class where you write a GetValues() function that provides the data you want to test. Can you elaborate?

**Comment by jfinkhaeuser** [comment]
It's a reporting thing more than a functionality thing. My example was not syntactically clear enough, really.

Imagine that the `foo` above is something like this:
```c++
struct foo
{
  int i;
};
```

And in one test case I want to use `foo{1}`, in the other `foo{2}`.

From a reporting point of view, I want to group all tests of the `foo` struct together. From a coverage point of view, I want to test a bunch of `foo` instances, assuming different values of `i` should yield different results. Currently, data-driven tests give me both.

What they don't give me is a way to add a test case that uses all or many of the test values to the same reporting group. That is, if I do want to add a test that all the different `foo` instances hash to unique values, that test would have to live outside of the `foo` reporting group.

The TL;DR is, what good is test group naming (via classes) if you can't group all related tests together?
